### PR TITLE
Adding entity uuid index

### PR DESCRIPTION
**Pull request by timmillwood**

### CODE REVIEW

**Comment by timmillwood** [comment]
https://www.drupal.org/node/2690745

**Comment by timmillwood** [comment]
@bojanz @dawehner Would love to hear thoughts / feedback etc.

**Comment by Berdir** [comment]
I was expecting that this would be an index to look up the entity for an UUID. Which would be useful because loadEntityByUuid() in core is very slow and a performance problem, especially for content blocks.

But this seems to be the opposite, I'm not really sure what that's good for?

If anything, I think we should support both, which means a custom table/storage, not key_value.

**Comment by timmillwood** [comment]
@Berdir it is to look up the entity for a UUID.

```
\Drupal::service('entity.index.uuid')->get($uuid);
```

This returns an array of `entity_type_id` and `entity_id` so you can get the entity. A more functional example would be:

```
$uuid = '558bdf0d-66fa-4f52-b708-152d65245beb';
$index = \Drupal::service('entity.index.uuid')->get($uuid);
$entity = \Drupal::entityTypeManager()->getStorage($index['entity_type_id'])->load($index['entity_id']);
```

**Comment by Berdir** [comment]
That's not what I'm seeing in the code?

You're talking about $key for the get methods, and buildKey() is $entity_type_id:$entity_id?

**Comment by Berdir** [comment]
hook_install() is also not going to scale like that :) you'd need to manually initialize it for existing sites, at least if the entity count is > 50 or so.

**Comment by timmillwood** [comment]
`EntityIndexBase` uses `$entity_type_id:$entity_id` as the key, then `UuidIndex` extends that using uuid as the key. This has been built to be flexible enough to setup varying indexes. For example once we get a revision uuid in core we could have `RevisionUuidIndex` which returns the entity for a revision uuid.

I just did a quick performance test, even enabling via the UI rather than Drush, it generated 1108 index entries in a split second.

**Comment by timmillwood** [comment]
It's been a couple of weeks since the last comment I just wanted to see if there was any interest in this?

File: src/Entity/Index/EntityIndexBase.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\KeyValueStore\KeyValueFactoryInterface;
+
+class EntityIndexBase implements EntityIndexInterface {
+
+  /**
+   * @var string
+   */
+  protected $collection = 'entity.index.id';
```
Comment (dawehner) [inline src/Entity/Index/EntityIndexBase.php:13]: what about using a constant here?

File: src/Entity/Index/EntityIndexBase.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      $this->cache[$key] = $value;
+    }
+    $this->keyValueStore->setMultiple($values);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function buildKey(EntityInterface $entity) {
+    return $entity->getEntityTypeId() . ':' . $entity->id();
```
Comment (dawehner) [inline src/Entity/Index/EntityIndexBase.php:102]: Mh, could we build it to have revision support in the future, maybe?

File: src/Entity/Index/EntityIndexInterface.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+use Drupal\Core\Entity\EntityInterface;
+
+interface EntityIndexInterface {
+
+  /**
+   * @param string $key
+   *
+   * @return array
+   */
+  public function get($key);
```
Comment (dawehner) [inline src/Entity/Index/EntityIndexInterface.php:14]: It is not obvious what $key is here

File: tests/src/Kernel/Entity/Index/UuidIndexTest.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace Drupal\Tests\entity\Kernal\Entity\Index;
```
Comment (dawehner) [inline tests/src/Kernel/Entity/Index/UuidIndexTest.php:3]: Should be Kernel instead of `Kernal`

File: src/Entity/Index/EntityIndexBase.php (hunk 1 above)
Comment (timmillwood) [inline src/Entity/Index/EntityIndexBase.php:13]: Yeh, that could work.

File: src/Entity/Index/EntityIndexBase.php (hunk 2 above)
Comment (timmillwood) [inline src/Entity/Index/EntityIndexBase.php:102]: In `buildValue()` we could store the latest / relevant revision id, we could also extent `EntityIndexBase` and create `RevisionIndex` to have an index based of revision uuids.

File: src/Entity/Index/EntityIndexInterface.php (hunk 3 above)
Comment (timmillwood) [inline src/Entity/Index/EntityIndexInterface.php:14]: In `EntityIndexBase` the key is entity type id followed by the entity id, for example `node:1`, this is pretty useless, but anything that extends this like `UuidIndex` does can set what their own key would be. `UuidIndex` uses the uuid.

File: src/Entity/Index/EntityIndexBase.php (hunk 2 above)
Comment (dawehner) [inline src/Entity/Index/EntityIndexBase.php:102]: Ah nice, yeah maybe a follow PR would be great!

File: src/Entity/Index/EntityIndexInterface.php (hunk 3 above)
Comment (dawehner) [inline src/Entity/Index/EntityIndexInterface.php:14]: well, I was more talking about the documentation to be hoenst

**Comment by timmillwood** [comment]
In Multiversion module I am currently working on some changes to the indexes, which will heavily influence how I'd want them to work. Therefore I am going to close this PR for now.

I'd like to open discussion about how this could work in the core issue https://www.drupal.org/node/2690747

**Pull request closed without merging**